### PR TITLE
Add more vesting and transactions data to grants export

### PR DESCRIPTION
**Pull request by 1emu**
- Add one time payment transaction date and amount
- Add vesting contract start date
- Appended `vesting_` to all vesting related fields
  
>   { id: 'vesting_address', title: 'Vesting Contract' },
>     { id: 'vesting_released', title: 'Vesting Released Amount' },
>     { id: 'vesting_releasable', title: 'Vesting Releasable Amount' },
>     { id: 'vesting_start_at', title: 'Vesting Start At' },
>     { id: 'vesting_finish_at', title: 'Vesting Finish At' },

image

### CODE REVIEW

File: src/export-transactions.ts [hunk 1]
```diff
@@ -77,17 +76,17 @@ const FACILITATOR_ADDRESS = '0x76fb13f00cdbdd5eac8e2664cf14be791af87cb0'
...
   console.log('Latest', network, block, latestBlock, (latestBlock - block) / 1000000)
 
   while (block < latestBlock) {
-    url = `https://api.covalenthq.com/v1/${network}/events/topics/${topic}/?key=${API_KEY}&starting-block=${block}&ending-block=${block + 1000000}&page-size=1000000000`
+    url = `https://api.covalenthq.com/v1/${network}/events/topics/${topic}/?key=${COVALENT_API_KEY}&starting-block=${block}&ending-block=${block + 1000000}&page-size=1000000000`
     console.log('fetch', url)
     json = await fetchURL(url)
     const data: APIEvents = json.data
-    let evs = data.items.map(e => e.tx_hash)
+    const evs = data.items.map(e => e.tx_hash)
```
Comment (andyesp) [inline src/export-transactions.ts:89]: I know it is not unrelated with your changes but... What is `evs`? Is it events? Maybe `eventsWithTxHash`?

File: src/export-grants.ts [hunk 2]
```diff
@@ -37,44 +42,93 @@ function parseNumber(n: number, decimals: number) {
...
+      const decodedLogEvent = logEvent.decoded
+      if (decodedLogEvent) {
+        if (decodedLogEvent.name === 'Transfer') {
+          const txMatchesBeneficiary = decodedLogEvent.params.some(param =>
+              param.name === ParamName.To && String(param.value).toLowerCase() === proposal.grant_beneficiary.toLowerCase());
+          if (txMatchesBeneficiary) {
+            proposal.token = DECIMALS[logEvent.sender_address][0]
+            const decimals: number = DECIMALS[logEvent.sender_address][1]
+            proposal.tx_date = logEvent.block_signed_at
+            findTxAmount(decodedLogEvent, proposal, decimals);
```
Comment (andyesp) [inline src/export-grants.ts:97]: I'd change this to...
```
proposal.tx_amount = getTxAmount(decodedLogEvent, decimals)
```

File: src/export-grants.ts [hunk 3]
```diff
@@ -37,44 +42,93 @@ function parseNumber(n: number, decimals: number) {
...
+    if(proposal.vesting_address === null && proposal.enacting_tx === null){
+      console.log(`A proposal without vesting address and enacting tx has been found. Id ${proposal.id}`)
     }
   }
 
   console.log(proposals.length, 'grants found.')
 
   saveToJSON('grants.json', proposals)
-  saveToCSV('grants.csv', proposals, [
+  await saveToCSV('grants.csv', proposals, [
```
Comment (andyesp) [inline src/export-grants.ts:131]: Should we await this?

File: src/export-grants.ts [hunk 4]
```diff
@@ -37,44 +42,93 @@ function parseNumber(n: number, decimals: number) {
...
+    proposal.grant_category = proposal.configuration.category
+    proposal.grant_tier = proposal.configuration.tier.split(':')[0]
+    proposal.grant_size = proposal.configuration.size
+    proposal.grant_beneficiary = proposal.configuration.beneficiary
 
-      } catch (e){
-        console.log(`Error trying to get vesting data for proposal ${p.id}`, e)
-      }
+    if (proposal.vesting_address) {
+      await fillVestingContractData(proposal);
```
[the remark this answers is not in the document]
Comment (1emu) [reply src/export-grants.ts]: TeenageUnmutableNinjaTurtles FTW

File: src/export-grants.ts (hunk 3 above)
Comment (1emu) [reply src/export-grants.ts:131]: Yes, it's an async method so we want to know that it finishes properly before returning

**Review by ncomerci — APPROVED** [review]

**Pull request merged**